Essentials/src/com/earth2me/essentials/Teleport.java

```diff
@@ -121,28 +121,38 @@ public void now(Player entity, boolean cooldown, TeleportCause cause) throws Exc
     protected void now(IUser teleportee, ITarget target, TeleportCause cause) throws Exception {
         cancel(false);
         teleportee.setLastLocation();
-        Location loc = target.getLocation();
 
-        if (LocationUtil.isBlockUnsafeForUser(teleportee, loc.getWorld(), loc.getBlockX(), loc.getBlockY(), loc.getBlockZ())) {
-            if (ess.getSettings().isTeleportSafetyEnabled()) {
-                if (ess.getSettings().isForceDisableTeleportSafety()) {
-                    PaperLib.teleportAsync(teleportee.getBase(), loc, cause);
+        PaperLib.getChunkAtAsync(target.getLocation()).thenAccept(chunk -> {
+            Location loc = target.getLocation();
+            if (LocationUtil.isBlockUnsafeForUser(teleportee, chunk, loc.getBlockX(), loc.getBlockY(), loc.getBlockZ())) {
+                if (ess.getSettings().isTeleportSafetyEnabled()) {
+                    if (ess.getSettings().isForceDisableTeleportSafety()) {
+                        PaperLib.teleportAsync(teleportee.getBase(), loc, cause);
```

Comment:
We never need an 'async' call inside any other 'async' callbacks because the chunk has already been loaded - this should make the code a lot nicer.

Reply:
This would be a single line change but sure I guess